tools/image-builder/main.tf

```diff
@@ -0,0 +1,98 @@
+module "landing_zone" {
+  source                                 = "terraform-ibm-modules/landing-zone/ibm"
+  version                                = "5.27.0"
+  prefix                                 = local.prefix
+  region                                 = local.region
+  tags                                   = local.tags
+  resource_groups                        = local.env.resource_groups
+  network_cidr                           = local.env.network_cidr
+  vpcs                                   = local.env.vpcs
+  vpn_gateways                           = local.env.vpn_gateways
+  enable_transit_gateway                 = local.env.enable_transit_gateway
+  transit_gateway_resource_group         = local.env.transit_gateway_resource_group
+  transit_gateway_connections            = local.env.transit_gateway_connections
+  ssh_keys                               = local.env.ssh_keys
+  vsi                                    = local.env.vsi
+  security_groups                        = local.env.security_groups
+  virtual_private_endpoints              = local.env.virtual_private_endpoints
+  cos                                    = local.env.cos
+  service_endpoints                      = local.env.service_endpoints
+  key_management                         = local.env.key_management
+  skip_kms_block_storage_s2s_auth_policy = local.env.skip_kms_block_storage_s2s_auth_policy
+  atracker                               = local.env.atracker
+  clusters                               = local.env.clusters
+  wait_till                              = local.env.wait_till
+  f5_vsi                                 = local.env.f5_vsi
+  f5_template_data                       = local.env.f5_template_data
+  appid                                  = local.env.appid
+  teleport_config_data                   = local.env.teleport_config_data
+  teleport_vsi                           = local.env.teleport_vsi
+}
+
+# # Code for Public Gateway attachment for the existing vpc and new subnets scenario
+
+data "ibm_is_public_gateways" "public_gateways" {
+}
+
+locals {
+  public_gateways_list = data.ibm_is_public_gateways.public_gateways.public_gateways
+  zone_1_pgw_ids       = var.vpc_name != null ? [for gateway in local.public_gateways_list : gateway.id if gateway.vpc == local.vpc_id && gateway.zone == var.zones[0]] : []
+}
+
+resource "ibm_is_subnet_public_gateway_attachment" "zone_1_attachment" {
+  count          = (var.vpc_name != null && var.subnet_id == null) ? 1 : 0
+  subnet         = module.landing_zone.subnet_data[0].id
+  public_gateway = length(local.zone_1_pgw_ids) > 0 ? local.zone_1_pgw_ids[0] : ""
+}
+
+resource "null_resource" "compress_and_encode_folder" {
```

Comment:
This whole section of `null_resources` and the data lookup worries me. It might be fine and may be your only option but I want to think about it more and fully understand what is trying to be accomplished.

Reply:
Actually we wanted to copy few files to remote VSI and we could'nt use remote-exec scp way because we want to give users an option to disable floating-ip as well and this scp approach doesn't work wihout fip.

So we thought about this solution here we dont even need the floating ip neither the scp and we can directly compress the files, encode and send the encoded files to remote VSI via userdata and then unpack the contents of compressed file under userdata so that the files can be used there normally. Hope I am able to explain.

Reply:
For any configuration on the VSI itself, Ansible would be preferred. Terraform is not well suited to this approach, I think maintenance of this will be a pain.  See how the PowerVS module handles this https://github.com/terraform-ibm-modules/terraform-ibm-powervs-infrastructure/blob/f728b65cf2b34b0660b3e00544efb306744a0d4f/modules/powervs-vpc-landing-zone/main.tf#L87

Reply:
Right but for ansible also we need either floating IP attached to the remote server or bastion in the middle. And in our scenario we want to have a solution which can copy files from local to remote machine without the mentioned ones.

Reply:
These are only temporary infrastructure resources to last hours/days no maintenance is expected. As such happy to approve this approach.